### ISSUE TITLE
[mlir] Marking analysis retained external to pass

### ISSUE THREAD

**Issue by jpienaar** [llvm/llvm-project#54178]
Currently there doesn't seem to be a way to mark an analysis as retained except by a pass author. That means if I'm an Analysis author I cannot mark an Analysis as retained for passes which I know it won't be affected/will be retained by but require the pass knows about the analysis and marks it retained. I know this is semi WAI as the set retained is a guarantee by pass author as currently written. But for cases where it is indirect, this results in conservative invalidations which cannot be avoided (e.g., canonicalization requires not changing type/inserting casts as needed, analysis Foo is valid while types are unchanged and so can depend on the invariant provided by canonicalization, but Analysis cannot be retained today unless we mark it in the pass itself - in this case an alternative could be a query method "is valid while: type_unchanged" and canonicalization could have "invariant: type_unchanged" and using these "traits" one could determine it was retained even without Foo being known to pass author; this shifts knowledge of these "traits"/invariants to pass and analysis authors in decoupling).

**Comment by River707** [comment]
Yeah, this is definitely an underdeveloped area of the infra. The only thing that exists in tree, is that an analysis can provide a `bool isInvalidated(PreservedAnalyses &pa)` that returns if the analysis was actually invalidated. It doesn't provide the operation the analysis was constructed on though.
